Gutenpack Subscription Block

gititon opened the pull request:
This pull request adds the Gutenberg Subscription block.

The Subscription block invites visitors to sign up to be notified when there are new posts on a site.

When the block has focus in the editor, editors can toggle whether to show the number of subscribers in the block. When it doesn't have focus, editors will see the number of subscribers if they toggled that on.

#### Testing instructions:
Note that the block still is going to receive some CSS ❤️(cc @MichaelArestad) , so please focus on testing its functionality for now.

1) Checkout this Jetpack branch and the corresponding Calypso branch, 
https://github.com/Automattic/wp-calypso/pull/28887

2) Run the command to build the Subscription block from your wp-calypso repo directory:

```
npm run sdk gutenberg client/gutenberg/extensions/presets/jetpack -- --output-dir=/<path_to_jetpack_repo>/_inc/blocks -w
```
3) Connect Jetpack if it isn't connected

4) Create a new post and add the Jetpack Subscription Block.

5) Toggle the "show number of subscribers". Confirm that you see the number of subscribers when the block loses focus.

6) Publish the post. Confirm that you see it on the frontend. 

7) Subscribe by using the block with a user or two (they will have to be a connected Jetpack user)

8) Confirm that you see the number of subscribed users on the frontend (when the toggle was enabled for the block in the editor) and the editor when the block doesn't have focus.

#### Proposed changelog entry for your changes:

No changelog entry needed.

gititon committed:
Changes for subscription block

gititon committed:
Changes for subscription block

gititon committed:
Update

gititon committed:
Update

jeherve commented on class.jetpack-gutenberg.php:
```diff
@@ -206,6 +206,14 @@ public static function enqueue_block_editor_assets() {
+		$is_subscriptions_active = Jetpack::is_module_active( 'subscriptions' );
+		if ( $is_subscriptions_active ) {
+			// Get number of subscribers
+			$subscriptions = new Jetpack_Subscriptions_Widget();
```
I'm getting the following error when the Subscriptions module is not active on the site:

`Fatal error: Class 'Jetpack_Subscriptions_Widget' not found in wp-content/plugins/jetpack-dev/_inc/lib/admin-pages/class.jetpack-react-page.php on line 235`

gititon replied:
Hey @jeherve , can you please tell me how I can reproduce this fatal error? I’ve tried disabling the Subscriptions module and editing a post with a Subscription block in it as well as creating a new post, but neither resulted in any errors in my error log.

jeherve commented on modules/subscriptions.php:
```diff
@@ -852,6 +852,11 @@ function widget( $args, $instance ) {
 		$show_subscribers_total = (bool) $instance['show_subscribers_total'];
 		$subscribers_total      = $this->fetch_subscriber_count(); // Only used for the shortcode [total-subscribers]
 
+		if ( $instance['show_only_email_and_button'] ) {
```
Why do we want to remove that information from the block? Wouldn't it be nice if folks could opt to use or customize that information?

gititon replied:
This allows the published post/form to conform to @MichaelArestad 's design, https://wp.me/pafL3P-4w-p2

The idea was that anyone who wants to add things to their subscription form could simply add other block types above, below, to the side of it, etc. and include anything they want.

jeherve commented on modules/subscriptions.php:
```diff
@@ -1122,3 +1130,10 @@ function jetpack_do_subscription_form( $instance ) {
+
+jetpack_register_block( 'subscriptions', array( 'render_callback' => 'show_subscription_block' ) );
+
+function show_subscription_block( $attr, $content ) {
```
I would recommend that you prefix that function to avoid potential conflicts with other plugins and themes.

jeherve reviewed (changes requested):
Switching this back to "In Progress", since the block is broken and some things will need to be updated in Jetpack to make it work:
https://github.com/Automattic/wp-calypso/pull/28366#pullrequestreview-176887677

This PR will also need a rebase.

ockham committed:
REST API: G'berg Available Extensions: Set Site Specific Endpoint Flag (#10740)

* REST API: G'berg Available Extensions: Factor out rest_base

* REST API: G'berg Available Extensions: Set Site Specific Endpoint Flag

simison committed:
Tiled gallery block: load assets (#10728)

Load Tiled Gallery Gutenberg block assets.

## Testing

- Apply the commit https://github.com/Automattic/jetpack/pull/10739/commits/9e27e5978eab605c6b06f7ad38390c01c69009a3 to fix loading block assets at appropiate loading order. (PR https://github.com/Automattic/jetpack/pull/10739)
- In Calypso, switch to branch `update/tiled-gallery-updates` (PR https://github.com/Automattic/wp-calypso/pull/27458) and build the blocks to get `_inc/blocks/tiled-gallery/view*` files:
    ```
    npm run sdk -- gutenberg client/gutenberg/extensions/presets/jetpack/ \
        --output-dir=~/jetpack/_inc/blocks/
    ```
- Insert tiled gallery block in the editor, add some awesome images and publish the post
- Check the post frontend, you should observe `tiled-gallery/view.*` files load and the gallery shouldn't look too broken. It might look just a tad bit broken because it's work in progress. ;-)

gititon committed:
Add Subscriber API Endpoint (#10741)

enejb committed:
Maps Block: Add an new api endpoint  (#10742)

* Add an new api endpoint

* Updates to match wpcom side.

* Added deprecated fucntion calls.

* Removed todo comments

* Updated to use the __METHOD__ magics

* Added the @deprecated tag.

* Add tests fot the api endpoint

gititon committed:
fixing call to is_module_active

gititon committed:
fixing call to is_module_active

gititon committed:
prefixing show_subscription_block() function to avoid conflicts

enejb commented on class.jetpack-gutenberg.php:
```diff
@@ -334,6 +334,14 @@ public static function enqueue_block_editor_assets() {
+		if ( $is_subscriptions_active ) {
+			// Get number of subscribers
+			$subscriptions = new Jetpack_Subscriptions_Widget();
+			$subscriber_info = $subscriptions->fetch_subscriber_count();
```
Does it make sense to convert fetch_subscriber_count to a static method so that we can just call it directly?

lezama replied:
good call, but the constructor does some magic, like adding xmlrpc endpoints that might be needed 🤔 

but this make me realize that we should call `Jetpack_Subscriptions_Widget::init()` instead of creating a new instance

gititon committed:
Making fetch_subscriber_count() a static method

gititon committed:
Updating increment_subscriber_count to behave correctly going from 0 to 1 subscribers

gititon committed:
Reverting change to increment count mechanism and deleting transient in endpoint so it's fetched fresh

gititon committed:
Wrapping subscriber endpoint transient deletion if ! testing conditional

lezama commented on _inc/lib/core-api/wpcom-endpoints/subscribers.php:
```diff
@@ -40,8 +40,12 @@ public function readable_permission_check() {
-		$subscriptions = new Jetpack_Subscriptions_Widget();
-		$subscriber_info = $subscriptions->fetch_subscriber_count();
+		// Get the most up to date subscriber count when request is not a test
+		if ( ! defined( 'TESTING_IN_JETPACK' ) ) {
```
does Jetpack side knows when the site gets a new subscriber? could we delete the transient at that point?

gititon replied:
@lezama I don't believe it does. There is an `increment_subscriber_count` function, https://github.com/Automattic/jetpack/blob/d7a1fadb4163d62d388c1b636c05f261fbab31fd/modules/subscriptions.php#L984, but I searched the Jetpack repo and don't see it invoked anywhere. 

I believe the way it works is that subscribers confirm their subscription via the email they receive from wpcom, and the Jetpack site simply drops its transient and queries wpcom if the transient is more than an hour old to get the latest confirmed subscriber count.

jeherve commented on modules/subscriptions.php:
```diff
@@ -850,7 +850,12 @@ function widget( $args, $instance ) {
 
 		$show_subscribers_total = (bool) $instance['show_subscribers_total'];
-		$subscribers_total      = $this->fetch_subscriber_count(); // Only used for the shortcode [total-subscribers]
+		$subscribers_total      = self::fetch_subscriber_count(); // Only used for the shortcode [total-subscribers]
```
It may be good to update the comment here now that it will also be used for the block?

jeherve commented on modules/subscriptions.php:
```diff
@@ -850,7 +850,12 @@ function widget( $args, $instance ) {
+		$subscribers_total      = self::fetch_subscriber_count(); // Only used for the shortcode [total-subscribers]
+
+		if ( $instance['show_only_email_and_button'] ) {
+			unset( $instance['title']);
```
You seem to be missing a space here.

jeherve reviewed (approved):
This looks good and works well. I only have a few minor comments.

I also do have a few remarks with things I expected when I first loaded the block; I don't know if those are blockers, but I think they're things that would be worth fixing before the release.

-------

That's an issue with the shortcode as well, but I thought I'd mention it as it can be a pretty annoying user experience:
- When you use the subscription widget, we have a container with a unique ID around the form, so when you subscribe and the page reloads, we can automatically scroll down to where the form is. 
- The shortcode and the block don't have such a unique ID, so when you subscribe, the page reloads and you may not know if the subscription was actually successful or what to do next.

---------

Jetpack's Contact Form block has taught me to fill in the forms inside the block editor, to create my own placeholder text in the form:
![image](https://user-images.githubusercontent.com/426388/49596339-5f992b00-f97a-11e8-9e14-f6173321b6a9.png)
![image](https://user-images.githubusercontent.com/426388/49596461-aab33e00-f97a-11e8-9726-0d5adf2a00b6.png)

That's consequently the first thing I did when I inserted the subscription form block, and it let me do it:

![image](https://user-images.githubusercontent.com/426388/49596522-cd455700-f97a-11e8-82b3-c1bce6af6a59.png)

I think we should have a consistent experience across blocks.

---------

Similar to above, I expected an option to change the "Subscribe" button text via an option in the inspector controls. 

---------

There is no HTML content saved in the editor apart from the Gutenberg comments:

```
<!-- wp:jetpack/subscriptions {"subscriber_count_string":"Join 2 other subscribers","show_subscribers_total":true} /-->
```

I think it would make sense to have some HTML saved there. We don't have to save the actual form's HTML output, but we could save the subscription shortcode, like we do with the Simple Payments block for example:

```
<!-- wp:jetpack/simple-payments {"content":"The crown of the Rebel Alliance fleet","email":"jeremy@example.com","price":43,"productId":298,"title":"X-Wing Fighter"} -->
[simple-payment id="298"]
<!-- /wp:jetpack/simple-payments -->
```

In our case, that would show us something like:

```
<!-- wp:jetpack/subscriptions {"subscriber_count_string":"Join 2 other subscribers","show_subscribers_total":true} -->
[jetpack_subscription_form title="" subscribe_text="" show_subscribers_total="1"]
<!-- /wp:jetpack/subscriptions -->
```

jeherve reviewed (dismissed):
Looking at my logs, I got the following notice while testing:

```
PHP Notice:  Undefined index: show_only_email_and_button in /var/www/html/wp-content/plugins/jetpack/modules/subscriptions.php on line 855
```

MichaelArestad commented:
@jeherve I agree with your suggestions around the Subscription form label. I don't think it is something we should fix for this release, but is something we should change for all Subscription forms. Here's the Issue I created elaborating on it: https://github.com/Automattic/wp-calypso/issues/29189

gititon committed:
Adding missing whitespace, adding isset check for  ['show_only_email_and_button'], removing outdated comment

jeherve commented:
This will now need a rebase, because of the changes mentioned in #10860, that cause the tests to fail.

gititon commented:
@jere

> This will now need a rebase, because of the changes mentioned in #10860, that cause the tests to fail.

@jeherve The branch has been rebased. Are there any issues or is the branch good to go? Thanks!

gititon committed:
Merge branch 'master' into gutenpack-subscriptions

roccotripaldi commented on modules/subscriptions.php:
```diff
@@ -1122,3 +1130,10 @@ function jetpack_do_subscription_form( $instance ) {
 }
+
+jetpack_register_block( 'subscriptions', array( 'render_callback' => 'jetpack_show_subscription_block' ) );
+
```
this needs to be registered in a file that is shared with .com

roccotripaldi replied:
actually, scratch that! there is not subscription-related file that is shared with .com

enejb commented on modules/subscriptions.php:
```diff
@@ -1122,3 +1130,10 @@ function jetpack_do_subscription_form( $instance ) {
+jetpack_register_block( 'subscriptions', array( 'render_callback' => 'jetpack_show_subscription_block' ) );
+
+function jetpack_show_subscription_block( $attr, $content ) {
+	$attr['show_only_email_and_button'] = true;
```
Instead of doing this can we just pass this attribute via the block save method?

enejb commented on _inc/lib/core-api/wpcom-endpoints/subscribers.php:
```diff
@@ -40,8 +40,12 @@ public function readable_permission_check() {
-		$subscriptions = new Jetpack_Subscriptions_Widget();
-		$subscriber_info = $subscriptions->fetch_subscriber_count();
+		// Get the most up to date subscriber count when request is not a test
+		if ( ! defined( 'TESTING_IN_JETPACK' ) ) {
```
Lets use Jetpack_Constant class here instead.

enejb commented on _inc/lib/core-api/wpcom-endpoints/subscribers.php:
```diff
@@ -50,6 +54,6 @@ public function get_subscriber_count( $request ) {
 }
 
-if ( Jetpack::is_module_active( 'subscriptions ') || ( defined( 'TESTING_IN_JETPACK' ) && TESTING_IN_JETPACK ) ) {
+if ( Jetpack::is_module_active( 'subscriptions' ) || ( defined( 'TESTING_IN_JETPACK' ) && TESTING_IN_JETPACK ) ) {
```
lets use `Jetpack_Constsant class here as well.

gititon committed:
Using Jetpack_Constants to test whether JETPACK_TESTING is defined instead of defined()

gititon committed:
Removing subscriptions render call back since the shortcode provided by save is sufficient

roccotripaldi reviewed (approved):
Tests well!

oskosk commented:
After following the testing instructions I went and deactivated subscriptions. Then reloaded the frontend for the post I just wrote. I see this

![image](https://user-images.githubusercontent.com/746152/50003947-35e89100-ff84-11e8-8e8b-304ced33a1fe.png)


On the editor, I refresh the page and I see this

![image](https://user-images.githubusercontent.com/746152/50003973-4d277e80-ff84-11e8-9bae-bb7b230f2c35.png)


I clicked Keep as html

![image](https://user-images.githubusercontent.com/746152/50004016-7ba55980-ff84-11e8-8412-62382d0bff72.png)

Then saved the post and visitted the frontend again

![image](https://user-images.githubusercontent.com/746152/50003947-35e89100-ff84-11e8-8e8b-304ced33a1fe.png)
(Same as when I just deactivated subscriptions).

Then I activated subscriptions again and refreshed the post editor page.
The block is rendered like this on the editor.

![image](https://user-images.githubusercontent.com/746152/50004122-d5a61f00-ff84-11e8-979e-e33c48c65939.png)

![image](https://user-images.githubusercontent.com/746152/50004128-de96f080-ff84-11e8-85fb-e31112d24fd9.png)


Then I updated the post, thinking maybe something different will happen but it didn't. Seems like I cannot recover the regular block rendering now.

That said, everything that was stated in the testing instructions worked. The subscriptions count was there.
